### PR TITLE
[node]: Update assert.fail parameters

### PR DESCRIPTION
**Pull request by eyqs**
[assert.fail](https://nodejs.org/api/assert.html#assert_assert_fail_message) has an optional `message` parameter that can be either a `string` or `Error`.

### CODE REVIEW

**Review by Flarna — CHANGES_REQUESTED** [review]
Could you please update also latest NodeJs (file types/node/index.d.ts).

Docs tell that `string | Error` is applicable to other assert apis also. Would be nice to update them also if you like.

**Comment by eyqs** [comment]
OK, I'm working on it.

**Comment by Flarna** [comment]
Changes in nodejs 10 look good.

For version 9.x I would recommend also `string | Error` (even docs say `any`) as implementation change to support `Error` in in since 9.0.0 (see https://github.com/nodejs/node/commit/e13d1df89bbaf26601b6c1c8406113b80bb6a95c).

I don't think the change from `string` to `any` for nodejs < 10.x should be done. I know the docs tell `any` but actually the source just passes the content of `message` to the `Error` constructor (except if it is `undefined`) which expects a `string`.

Type specification in nodejs docs is getting better and better but for old nodejs versions it's often to unspecific/sloppy for direct reuse here therefore I tend to take a look into the source code also.

File: types/node/v0/index.d.ts [hunk 1]
```diff
@@ -1867,32 +1867,27 @@ declare module "assert" {
...
-            constructor(options?: {message?: string; actual?: any; expected?: any;
-                                  operator?: string; stackStartFunction?: Function});
+            constructor(options?: {
+                message?: string; actual?: any; expected?: any;
+                operator?: string; stackStartFn?: Function
+            });
         }
 
-        export function fail(actual?: any, expected?: any, message?: string, operator?: string): never;
+        export function fail(actual: any, expected: any, message: any, operator: string): never;
```
Comment (Flarna) [inline types/node/v0/index.d.ts]: I think this change is not needed or an overload `fail(message?: string)` should be added.

File: types/node/v4/index.d.ts [hunk 2]
```diff
@@ -2449,31 +2449,29 @@ declare module "assert" {
             expected: any;
             operator: string;
             generatedMessage: boolean;
+            code: 'ERR_ASSERTION';
```
Comment (Flarna) [inline types/node/v4/index.d.ts]: I think this field is not yet present in node 4

File: types/node/v4/index.d.ts [hunk 3]
```diff
@@ -2449,31 +2449,29 @@ declare module "assert" {
...
-            constructor(options?: {message?: string; actual?: any; expected?: any;
-                                  operator?: string; stackStartFunction?: Function});
+            constructor(options?: {
+                message?: string; actual?: any; expected?: any;
+                operator?: string; stackStartFn?: Function
+            });
         }
 
-        export function fail(actual: any, expected: any, message?: string, operator?: string): never;
+        export function fail(actual: any, expected: any, message: any, operator: string): never;
```
Comment (Flarna) [inline types/node/v4/index.d.ts]: I think this change is not needed or an overload fail(message?: string) should be added.

File: types/node/v6/index.d.ts [hunk 4]
```diff
@@ -3941,33 +3941,30 @@ declare module "assert" {
             expected: any;
             operator: string;
             generatedMessage: boolean;
+            code: 'ERR_ASSERTION';
```
Comment (Flarna) [inline types/node/v6/index.d.ts]: this field is not yet present in node v6

File: types/node/v6/index.d.ts [hunk 5]
```diff
@@ -3941,33 +3941,30 @@ declare module "assert" {
...
 
             constructor(options?: {
                 message?: string; actual?: any; expected?: any;
-                operator?: string; stackStartFunction?: Function
+                operator?: string; stackStartFn?: Function
             });
         }
 
-        export function fail(actual: any, expected: any, message?: string, operator?: string): never;
+        export function fail(message: string): never;
```
Comment (Flarna) [inline types/node/v6/index.d.ts]: message is optional

File: types/node/v7/index.d.ts [hunk 6]
```diff
@@ -3985,33 +3985,29 @@ declare module "assert" {
...
 
             constructor(options?: {
                 message?: string; actual?: any; expected?: any;
-                operator?: string; stackStartFunction?: Function
+                operator?: string; stackStartFn?: Function
             });
         }
 
-        export function fail(actual: any, expected: any, message?: string, operator?: string): never;
+        export function fail(actual: any, expected: any, message: any, operator: string): never;
```
Comment (Flarna) [inline types/node/v7/index.d.ts]: fail without arguments is missing here, message should be an optional string

File: types/node/v8/index.d.ts [hunk 7]
```diff
@@ -5634,34 +5634,30 @@ declare module "assert" {
...
+            code: 'ERR_ASSERTION';
 
             constructor(options?: {
                 message?: string; actual?: any; expected?: any;
-                operator?: string; stackStartFunction?: Function
+                operator?: string; stackStartFn?: Function
             });
         }
 
         export function fail(message: string): never;
```
Comment (Flarna) [inline types/node/v8/index.d.ts]: message is optional

File: types/node/v9/index.d.ts [hunk 8]
```diff
@@ -5717,34 +5718,30 @@ declare module "assert" {
...
-        export function throws(block: Function, error: Function, message?: string): void;
-        export function throws(block: Function, error: RegExp, message?: string): void;
-        export function throws(block: Function, error: (err: any) => boolean, message?: string): void;
-
-        export function doesNotThrow(block: Function, message?: string): void;
-        export function doesNotThrow(block: Function, error: Function, message?: string): void;
-        export function doesNotThrow(block: Function, error: RegExp, message?: string): void;
-        export function doesNotThrow(block: Function, error: (err: any) => boolean, message?: string): void;
+        export function fail(message?: string | Error): never;
+        export function fail(actual: any, expected: any, message?: string | Error): never;
```
Comment (Flarna) [inline types/node/v9/index.d.ts]: Please add parameter `stackStartFn?: Function`.

**Comment by eyqs** [comment]
Sorry about all that, I was still basing it on the documentation. I'll revert the entire commit except the AssertionError part.

File: types/node/v9/index.d.ts [hunk 9]
```diff
@@ -5727,7 +5727,7 @@ declare module "assert" {
         }
 
         export function fail(message?: string | Error): never;
-        export function fail(actual: any, expected: any, message?: string | Error): never;
+        export function fail(actual: any, expected: any, message?: string, operator?: string, stackStartFn?: Function): never;
```
Comment (Flarna) [inline types/node/v9/index.d.ts]: `| Error` got lost here

File: types/node/v0/index.d.ts [hunk 10]
```diff
@@ -1868,31 +1868,25 @@ declare module "assert" {
...
-            constructor(options?: {message?: string; actual?: any; expected?: any;
-                                  operator?: string; stackStartFunction?: Function});
+            constructor(options?: {
+                message?: string; actual?: any; expected?: any;
+                operator?: string; stackStartFn?: Function
+            });
         }
 
-        export function fail(actual?: any, expected?: any, message?: string, operator?: string): never;
+        export function fail(actual?: any, expected?: any, message?: any, operator?: string): never;
```
Comment (Flarna) [inline types/node/v0/index.d.ts]: `any`=> `string` for message

File: types/node/v4/index.d.ts [hunk 11]
```diff
@@ -2450,30 +2450,27 @@ declare module "assert" {
...
-            constructor(options?: {message?: string; actual?: any; expected?: any;
-                                  operator?: string; stackStartFunction?: Function});
+            constructor(options?: {
+                message?: string; actual?: any; expected?: any;
+                operator?: string; stackStartFn?: Function
+            });
         }
 
-        export function fail(actual: any, expected: any, message?: string, operator?: string): never;
+        export function fail(actual: any, expected: any, message?: any, operator?: string): never;
```
Comment (Flarna) [inline types/node/v4/index.d.ts]: `any`=> `string` for message

File: types/node/v9/index.d.ts [hunk 12]
```diff
@@ -5717,34 +5718,30 @@ declare module "assert" {
...
-        export function throws(block: Function, error: Function, message?: string): void;
-        export function throws(block: Function, error: RegExp, message?: string): void;
-        export function throws(block: Function, error: (err: any) => boolean, message?: string): void;
-
-        export function doesNotThrow(block: Function, message?: string): void;
-        export function doesNotThrow(block: Function, error: Function, message?: string): void;
-        export function doesNotThrow(block: Function, error: RegExp, message?: string): void;
-        export function doesNotThrow(block: Function, error: (err: any) => boolean, message?: string): void;
+        export function fail(message?: string | Error): never;
+        export function fail(actual: any, expected: any, message?: string, operator?: string, stackStartFn?: Function): never;
```
Comment (Flarna) [inline types/node/v9/index.d.ts]: `| Error` got lost here

**Comment by eyqs** [comment]
Thank you for looking over the PR so thoroughly!

**Comment by eyqs** [comment]
I squashed the commits into one.

**Review by Flarna — APPROVED** [review]

**Comment by eyqs** [comment]
CI is still failing, was it OK to merge?

**Comment by andy-ms** [comment]
Yes, we're fixing the CI separately. It's not because of this change.

**Pull request merged**